File: android/src/main/java/app/independo/capacitorvoicerecorder/adapters/RecordDataMapper.java
package app.independo.capacitorvoicerecorder.adapters;

import app.independo.capacitorvoicerecorder.core.RecordData;
import com.getcapacitor.JSObject;

/** Maps internal record data into legacy or normalized JS payloads. */
public final class RecordDataMapper {

    private RecordDataMapper() {}

    /** Converts record data to the legacy payload shape. */
    public static JSObject toLegacyJSObject(RecordData recordData) {
        return recordData.toJSObject();
    }

    /** Converts record data to the normalized payload shape. */
    public static JSObject toNormalizedJSObject(RecordData recordData) {
        JSObject normalized = new JSObject();
        normalized.put("msDuration", recordData.getMsDuration());
        normalized.put("mimeType", recordData.getMimeType());

        String uri = recordData.getUri();
        String recordDataBase64 = recordData.getRecordDataBase64();
        if (uri != null && !uri.isEmpty()) {
            normalized.put("uri", uri);
        } else if (recordDataBase64 != null && !recordDataBase64.isEmpty()) {
            normalized.put("recordDataBase64", recordDataBase64);
        }

        return normalized;
    }
}

File: ios/Sources/VoiceRecorder/Adapters/RecordDataMapper.swift
import Foundation

/// Maps record data into legacy or normalized dictionary payloads.
struct RecordDataMapper {
    /// Converts record data to the legacy payload shape.
    static func toLegacyDictionary(_ recordData: RecordData) -> Dictionary<String, Any> {
        return recordData.toDictionary()
    }

    /// Converts record data to the normalized payload shape.
    static func toNormalizedDictionary(_ recordData: RecordData) -> Dictionary<String, Any> {
        var normalized: Dictionary<String, Any> = [
            "msDuration": recordData.msDuration,
            "mimeType": recordData.mimeType,
        ]

        if let uri = normalizedUri(from: recordData.uri) {
            normalized["uri"] = uri
        } else if let base64 = recordData.recordDataBase64, !base64.isEmpty {
            normalized["recordDataBase64"] = base64
        }

        return normalized
    }

    /// Normalizes legacy file paths into file:// URIs.
    private static func normalizedUri(from legacyUri: String?) -> String? {
        guard let legacyUri = legacyUri, !legacyUri.isEmpty else {
            return nil
        }

        if legacyUri.hasPrefix("file://") {
            return legacyUri
        }

        return URL(fileURLWithPath: legacyUri).absoluteString
    }
}
